File: LeetCodeProblemsGo/LeetCode744FindSmallestLetterGreaterThanTarget/LeetCode744FindSmallestLetterGreaterThanTarget.go
package LeetCode744FindSmallestLetterGreaterThanTarget

func nextGreatestLetter(letters []byte, target byte) byte {
	if target=='z'{
		return letters[0]
	}
 length:= len(letters)   
	begin:=0
	end:= length-1
	mid:=0
	found:= length
	for begin<=end{
		mid= (begin+end)<<1
		if letters[mid]> target{
			found= min(found,mid)
			end=mid-1
			continue
		}
		begin=mid+1
	}
	if found==length{
		return letters[0]
	}
	return letters[found]
}

File: LeetCodeProblemsCMake/Problems/LeetCode744FindSmallestLetterGreaterThanTarget.cpp
//
// Created by Saeed Jannati on 1/31/2026 A.
//

#include "LeetCode744FindSmallestLetterGreaterThanTarget.h"
using namespace std;

char LeetCode744FindSmallestLetterGreaterThanTarget::nextGreatestLetter(vector<char> &letters, char target) {
    if (target=='z') {
        return letters[0];
    }
    int length= static_cast<int>(letters.size());
    int begin{0};
    int end{length-1};
    int mid{0};
    int found{length};
    while (begin<=end) {
        mid=(begin+end)>>1;
        if (letters[mid]>target) {
            found=min(found,mid);
            end=mid-1;
            continue;
        }
        begin=mid+1;
    }
    if (found==length) {
        return letters[0];
    }
    return letters[found];
}
